google/cloud/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,275 @@
+using StreamReinitializer = std::function<future<StatusOr<
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>>(
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>)>;
+
+template <typename ResponseType, typename RequestType>
+class ResumableAsyncStreamingReadWriteRpc
+    : public AsyncStreamingReadWriteRpc<RequestType, ResponseType> {
+ public:
+  virtual future<Status> Status() = 0;
+};
+
+template <typename ResponseType, typename RequestType>
+class ResumableAsyncStreamingReadWriteRpcImpl
+    : public ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType> {
+ public:
+  ResumableAsyncStreamingReadWriteRpcImpl(
+      std::unique_ptr<RetryPolicy const> retry_policy,
+      std::unique_ptr<BackoffPolicy const> backoff_policy,
+      const AsyncSleeper& sleeper,
+      AsyncStreamFactory<ResponseType, RequestType>& stream_factory,
+      StreamReinitializer<ResponseType, RequestType>& reinitializer)
+      : retry_policy_prototype_(std::move(retry_policy)),
+        backoff_policy_prototype_(std::move(backoff_policy)),
+        sleeper_(std::move(sleeper)),
+        stream_factory_(std::move(stream_factory)),
+        reinitializer_(std::move(reinitializer)),
+        impl_(stream_factory_()) {}
+
+  ResumableAsyncStreamingReadWriteRpcImpl(ResumableStreamingReadRpc&&) = delete;
+  ResumableAsyncStreamingReadWriteRpcImpl& operator=(
+      ResumableStreamingReadRpc&&) = delete;
+
+  void Cancel() override {
+    // not guarding since this is the singular writer
+    is_canceled_ = true;
+    std::lock_guard<std::mutex> g{mu_};
+    impl_->Cancel();
+  }
+
+  future<bool> Start() override {
+    std::lock_guard<std::mutex> g{mu_};
+    return impl_->Start().then([this](future<bool> start_fu) {
+      if (!start_fu.get()) {
+        std::shared_ptr<RetryPolicy const> retry_policy =
+            this->retry_policy_prototype_->clone();
+        std::shared_ptr<BackoffPolicy const> backoff_policy backoff_policy =
+            this->backoff_policy_prototype_->clone();
+        return this->RetryLoop(retry_policy, backoff_policy, RetryType.Start);
+      }
+      return make_ready_future(true);
+    });
+  }
+
+  future<absl::optional<Response>> Read() override {
+    std::lock_guard<std::mutex> g{mu_};
+    if (!read_future_.has_value()) {
+      ProcessRead();
+    }
+    return read_future_.value();
+  }
+
+  future<bool> Write(Request const& r, grpc::WriteOptions o) override {
+    std::lock_guard<std::mutex> g{mu_};
+    if (!write_future_.has_value()) {
+      ProcessWrite(r, o);
+    }
+    return write_future_.value();
+  }
+
+  future<bool> WritesDone() override {
+    std::lock_guard<std::mutex> g{mu_};
+    return impl_->WritesDone().then([this](future<bool> writes_done_fu) {
+      bool writes_done = writes_done_fu.get();
+      std::lock_guard<std::mutex> g{this->mu_};
+      impl_->Finish().then([this](future<Status> finish_future) {
```

Comment:
Oh didn't realize that could happen. From Dan's comments, I was thinking of restructuring how and when `RetryLoop` is called and having a bit more granular use of mutexes. I'll keep this in mind when working on it, thanks!

Reply:
#7789 might serve as inspiration

Reply:
I thought about this more and have a few questions. I understand that if line 122 and the lambda within the `then` are on the same thread, then it's a deadlock since that thread is trying to acquire a lock it already has.
Though doesn't the continuation of the future occur on a separate 'executor' thread in which case we have a separate thread trying to acquire the lock? Or could it be the case that the 'executor' thread is the current thread?